### DISCUSSION TITLE
Build docker images for every master commit rather than just releases

### DISCUSSION THREAD

**Discussion by MOZGIII** [plantuml/plantuml#1131] [General]
I'd like to suggest that we build and push the docker image for each push to `master` instead of doing it only when the release is built.

The rationale for this is it will allow people to use the live master if they need to - for instance after they've contributed something, but before it could be released.

Tagging policy can be changed in the following way:
- each image gets tagged with `sha-<commit>` prefix (like we do it now)
- images built off of the commits to `master` are built with an additional `edge` tag
- images built off of git release tags get tagged with `latest` and `v<version>`

The Dockerfile will have to be altered in such a way that it doesn't depend on release - probably the easiest would be to just copy the `.jar` over from the `build` dir (`COPY /app/build/libs/plantuml-*-SNAPSHOT.jar /opt/plantuml.jar`). This will also eliminate the complexity of `wget`, and make the image-building process friendlier for local development and testing.

CC @gliptak

**Comment by gliptak** [comment]
to accomplish we might need to move `jar artifact` generation commands https://github.com/plantuml/plantuml/blob/master/.github/workflows/ci.yml#L128 into the `Dockerfile` to ensure single build path

**Reply by MOZGIII** [reply]
This is an option, but it will be doing a double job (on the build server) and, also, require duplicate efforts to maintain. It's way easier to just download the required `.jar` from the Github Actions artifacts (we can get them after the `upload` job).

How about that?

**Reply by gliptak** [reply]
so download or copy in the already built jar from the same run

https://github.com/plantuml/plantuml/pull/488

context mounts to root?

https://github.com/Devcon4/Picidae/blob/main/client/dockerfile.ci#L3

**Reply by MOZGIII** [reply]
We can control where we map the context with the `context` setting of the action.
Also, why not use root? It makes total sense, all we have to do is tweak the `.dockerignore` accordingly.

**Reply by MOZGIII** [reply]
> so download or copy in the already built jar from the same run

I'd straight up drop the download part - it will not be needed anymore, and it will take extra effort to keep around and maintain.

**Reply by matthew16550** [reply]
It would be nice to use `snapshot` as a tag to match the naming of jars and GH releases.

Is `edge` a commonly used tag?  I haven't been paying much attention to containers recently and don't remember seeing it.  No objection to using `edge` as well if that is the fashion though perhaps it adds confusion because PlantUML already uses the word "edge" for something else.

**Reply by arnaudroques** [reply]
This is not completely related, but since you guys are working on this, it would be **great** if we could also include publishing to maven central / OSSRH for each release.

This way, we could definitively move to Github and forget about sourceforge.net. Today, the "official" build for maven central is done by running this [pom.xml](https://sourceforge.net/p/plantuml/code/HEAD/tree/trunk/pom.xml), which is completely confusing since build on Github is done with Gradle :-) And this is the last reason why we keep sourceforge.

It should be easy to do, since `OSSRH_PASSWORD` and `OSSRH_USERNAME` are already present in the repository secrets. Unfortunately, I have no idea about how to set up this, so any help is welcome :-)

**Reply by MOZGIII** [reply]
`edge` is just one of the operational mode of the `docker/metadata-action`: https://github.com/docker/metadata-action#typeedge

Thus it is kind of an emerging standard - however we can use something other than the default `edge` tag (like `snapshot`) without any trouble.
